kvwrapper_etcd/kvwrapper_etcd.go

```diff
@@ -52,7 +67,11 @@ func (e EtcdWrapper) GetVal(key string) (*kvwrapper.KeyValue, error) {
 
 // GetVal returns a []KeyValue found at key
 func (e EtcdWrapper) GetList(key string, sort bool) ([]*kvwrapper.KeyValue, error) {
-	r, err := e.client.Get(key, sort, true)
+	options := &etcd.GetOptions{
+		Sort:      sort,
+		Recursive: true,
+	}
+	r, err := e.kapi.Get(context.Background(), key, options)
 	if err != nil {
 		if strings.HasPrefix(err.Error(), "501:") {
```

Comment:
Ah ok. So good thing you mentioned this as the error handling seems to be slightly different with the new lib.

The behavior when a key doesn't exist is the same; both libs return `500:` and so the code as-is here correctly handles this case.

However in the case of failing to connect to etcd, the old lib returned `100:` but the new lib instead returns an  OS-level error with a socket exception or such:

```
Get http://192.168.223.128:4001/v2/keys/flight-director/leader?quorum=false&recursive=true&sorted=false: dial tcp 192.168.223.128:4001: getsockopt: connection refused]) [0]
```

I think the easiest way to handle this is to change the code to treat `500:` as a missing key and everything else as `kvwrapper.ErrCouldNotConnect`, because it is likely that there are several different errors we can get depending on whether:

- no process is listening on the TCP port
- DNS lookup fails
- Out of file descriptors
etc...

It turns out we have some code in Flight Director that looks specifically for `kvwrapper.ErrCouldNotConnect` so this is the safest error to return, as it is most likely to make FD consider what happened to be a failure.

Implemented this in 3126060.